### ISSUE TITLE
[FEATURE-REQUEST] Could 'memory mapping' be 'imitated' with parquet files?

### ISSUE THREAD

**Issue by yohplala** [vaexio/vaex#1726]
**Description**
I have no real knowledge about memory mapping and probably what I am saying is much too naive.
Because data can be partitioned into row groups in parquet files, and that per row group, per column, statistics are available informing per row group and per column about
- min
-  max
- number of rows

I wonder if there could be a 'kind of' memory mapping done by vaex to load the data it needs (at row group level, not row level though)?

I believe to have noticed that vaex works by chunk of 1024 rows.
If I set this size for my row groups, could vaex take advantage of that to roll out a 'memory-mapping-like' approach?

**Is your feature request related to a problem? Additional context**
I have a use case for which multiple join operations have to be managed.
Basically, I am looking for joining 240 dataframes (from different parquet files) that contain each about 80 columns.
End dataframe will thus have about 19200 columns.
Joining will be managed on timestamps, but I do not necessarily need to ask for joining on timestamps because I know the same timestamps will be in all dataframe. I thus only need to select the data between 2 chosen timestamps, and join directly on rows.

Data is currently stored in parquet format.
I would like to keep this format because this format keeps size of data on disk small. This is convenient for the 'training phase' (lot of data on disk).

**Comment by JovanVeljanoski** [comment]
I didn't get the first part of the question, but if you want to just join directly on rows, you can do it right away by leaving the `on=` arguments empty. This is the fastest join vaex provides out of the box.

**Comment by yohplala** [comment]
> I didn't get the first part of the question, but if you want to just join directly on rows, you can do it right away by leaving the `on=` arguments empty. This is the fastest join vaex provides out of the box.

Hi @JovanVeljanoski 
Sorry, this ticket is related to ticket #1591. I should probably have continued there, but I feel this is approaching the topic by another direction.

Basically, in #1591, I report that joining over a LOT of dataframes which data is in parquet file push vaex to its limit (and actually, I used much less dataframes in the case I report in #1591 than the real targeted use case).
- Answer from @maartenbreddels was: "convert to arrow 1st"
- I translated into my mind as: "vaex needs memory mapped data to be able to process [such an amount of columns|such an amount of dataframes]"

Wondering if there is really no other way (because in arrow, data takes a lot of space), I am asking if 'somehow', by reducing size of row groups, vaex could not make use of min/max values per group to do 'as if' it could memory map parquet file content.

**Comment by maartenbreddels** [comment]
> I believe to have noticed that vaex works by chunk of 1024 rows.

Chunk size is dynamically between 1024 and 1024**2, see https://github.com/vaexio/vaex/blob/e2db5578d6805d7295a00d3440986d7622a473d8/packages/vaex-core/vaex/execution.py#L26

I think the issue still persists that the way we do join now requires random access into the dataframe that is joined (assuming a left join).
Random access is faster is the joined dataframe has data backing it that can be randomly accessed, the fastest being hdf5 (no chunks).

Alternatives are to materialise the full dataset backing the dataframe into memory using df.materialize (I think materialising datasets is not released yet). This should speed up the join at the cost of some memory.

The other alternative, is for join to not 'lazily' join, but eagerly materialise the data such that no random access is required. This is a feature that could be really useful for Vaex is situations like these (at the cost of even more memory).

**Comment by yohplala** [comment]
Hi @maartenbreddels 
Thanks a lot for taking the time to explain.
If I understand correctly, the 2 last solutions you present are the same, but one is activated by the user, whilee the other would be 'automatic', 'decided' by vaex on its own, right?

> Alternatives are to materialise the full dataset backing the dataframe into memory using df.materialize (I think materialising datasets is not released yet).

=> choice form the user

> The other alternative, is for join to not 'lazily' join, but eagerly materialise the data such that no random access is required. This is a feature that could be really useful for Vaex is situations like these (at the cost of even more memory).

=> decided by vaex

Thanks a lot for these feedbacks!
I will keep an eye on next changelogs to see when `df.materialize` is released.
Thanks again!

**Comment by yohplala** [comment]
Hi @maartenbreddels 
In the new FAQ item, for chosing appropriate file format, regarding Parquet, it is indicated (as per your advice above):
_Note that by using df.materialize one can get the same performance as HDF5 or Arrow files at the cost of memory or disk space._

In my case above (`join`), while a single dataframe fits in memory, the join result does not. And on this join result I am conducting additional vaex processing.

Please, how can you tell [`df.materialize`](https://vaex.io/docs/api.html#vaex.dataframe.DataFrame.materialize) to materialize the dataframe on disk, that in memory, it will not fit?
Or does vaex knows on its own?

Also, the docstring does not state it is possible to materialize on disk.

Thanks for your feedback!
Bests

**Comment by JovanVeljanoski** [comment]
So you need to know how join works i guess. 
At the moment, join takes memory, the cost of the column that serves as mapping from the left to right. So if your dataframe barely fits in memory, and you do a join (or more join) to other large dataframes, you can get out of memory. 
It is possible for us to do lazy join, but that is non trivial to implement and is subject to funding.

`.materialize` puts a column in memory. You don't "materialize a column to disk". If you have a virtual column that is very expensive to use on the fly, and you use it a lot, it might be better to have it in memory. Or if you want to have a part of a parquet file in memory instead of lazy reading it (which costs CPU due to the decompression) - the in-memory will be faster. 

If you are working with hdf5 or arrow, and the data is already on disk, materialize does not make much sense to use.

**Comment by yohplala** [comment]
@JovanVeljanoski  thanks for the answer.

Please, what is meant by "at the cost of disk space" in:
_Note that by using df.materialize one can get the same performance as HDF5 or Arrow files at the cost of memory or disk space._
?
How can we tell `materialize` to cache on disk? (this is my understanding)

> At the moment, join takes memory, the cost of the column that serves as mapping from the left to right. So if your dataframe barely fits in memory, and you do a join (or more join) to other large dataframes, you can get out of memory.
It is possible for us to do lazy join, but that is non trivial to implement and is subject to funding.

Yes, I understand. I have given thought about a possible implementation, and before actually coding, I am still enquiring about possible existing possibilities. :)

**Comment by JovanVeljanoski** [comment]
If you do materialize a column to parquet, and you export it as hdf5 or arrow, it will take more disk space, since parquet is compressed. 

You have to do things yourself -> materialize is just putting columns in memory. It is up to you to export to disk, in a format that is faster (arrow/hdf5).

**Comment by yohplala** [comment]
> You have to do things yourself -> materialize is just putting columns in memory. It is up to you to export to disk, in a format that is faster (arrow/hdf5).

Thanks a lot, crystal clear!
